Poll slurm to catch jobs that exited without writing output file

takluyver opened the pull request:
A partial solution for #18. As we discussed, polling Slurm is the crude but complete option: it should catch any kind of crash or cancellation, but we don't want to poll Slurm too often, so there's a significant lag before it picks up the failure. I've made the polling interval 30 seconds by default.

I've implemented this by extending the existing FileWaitThread to also poll Slurm on every 30th check. This has some limitations compared to running a separate thread - you can only poll Slurm just after the file checks, so checking the files every 10 seconds and Slurm jobs every 15 is not an option - but in general I think it's easier to follow code and avoid bugs if I don't use more threads than necessary. But I've ended up catching all exceptions from jobs_finished so they don't make it give up on checking files, which feels kind of kludgy. :confused: 

I've dropped Python 3.5 support in this PR. It's easy to add back if required, but even 3.6 is already EOL, and it's a bit of effort to remember not to use f-strings.

Tested by modifying the example:

```python
import cfut
import sys

def square(n):
    if n  == 7:
        sys.exit()
    return n * n

with cfut.SlurmExecutor(additional_setup_lines=['#SBATCH -p exfel']) as executor:
    for result in executor.map(square, [5, 7, 11]):
        print(result)
```

```
Traceback (most recent call last):
  File "/home/kluyvert/scratch/cfut-crash-worker.py", line 10, in <module>
    for result in executor.map(square, [5, 7, 11]):
  File "/gpfs/exfel/sw/software/xfel_anaconda3/1.1.2/lib/python3.7/concurrent/futures/_base.py", line 586, in result_iterator
    yield fs.pop().result()
  File "/gpfs/exfel/sw/software/xfel_anaconda3/1.1.2/lib/python3.7/concurrent/futures/_base.py", line 425, in result
    return self.__get_result()
  File "/gpfs/exfel/sw/software/xfel_anaconda3/1.1.2/lib/python3.7/concurrent/futures/_base.py", line 384, in __get_result
    raise self._exception
cfut.JobDied: Cluster job 10484726 finished without writing a result
```

takluyver committed:
Poll Slurm job status to catch jobs that finished without writing output file

takluyver committed:
Job IDs are stored as integers

takluyver committed:
Move thread shutdown check out of .check() method

takluyver committed:
Bump required Python to >= 3.6

takluyver committed:
Mock out squeue in Slurm tests

takluyver committed:
Mock jobs_finished function instead of squeue command

takluyver committed:
Catch exceptions from jobs_finished()

tcompa commented:
Hi there, I'm writing to ask for the status of this PR. It would be quite useful for us, and I'd be glad to contribute if this is helpful.

---

A bit of context:

As part of the [Fractal project](https://fractal-analytics-platform.github.io), we have a [server component](https://github.com/fractal-analytics-platform/fractal-server) which (also) handles the submission of SLURM jobs. We were using `parsl` in the past (via [SlurmProvider](https://parsl.readthedocs.io/en/stable/stubs/parsl.providers.SlurmProvider.html) and [HighThroughputExecutor](https://parsl.readthedocs.io/en/stable/stubs/parsl.executors.HighThroughputExecutor.html#parsl.executors.HighThroughputExecutor)) in the past, and we moved to `clusterfutures` for a few reasons (we were only using a tiny fraction of the `parsl` flexibility/features, and moreover it is very convenient for us to deal with standard `concurrent.futures` executors).

The issues in #19 are common to our use cases (see e.g. [this issue](https://github.com/fractal-analytics-platform/fractal-server/issues/343)): If a SLURM job does not correctly reach its end, we would like to see an error raised, so that our server can then capture it and let the user know.

More generally, we are finding `clusterfutures` quite useful, and we're heavily relying on it. We'd be glad to contribute back a bit, for instance starting with work on this PR.

takluyver commented:
Hi @tcompa ! I was waiting for @sampsyo to look at this, and then I forgot about it. He's given me permission to make changes and releases directly, but I'm used to working through pull requests and code reviews.

So perhaps a useful starting point would be for you - or someone in your team - to review and/or test my changes in this PR. If you're happy and @sampsyo is still busy with other things, I can merge it.

As discussed in #18, this is the slow but (hopefully) robust mechanism - it should eventually catch any job that fails to complete for any reason, but it may take up to 30 seconds to notice a failed job. I roughly planned to follow it with a faster mechanism for common cases (like jobs that are pre-empted or hit their time limit), but that gets more complex.

I also [tried to suggest](https://bugs.schedmd.com/show_bug.cgi?id=14638) that Slurm should have an efficient (non-polling) mechanism to wait for a job to finish, but so far didn't get any interest.

tcompa commented:
That's very good to know, thanks!
I will definitely have a look and test it a bit further, to see whether I spot anything weird or unexpected.

By now I just forked your branch, for testing, and implemented tiny changes (you can see them [here](https://github.com/takluyver/clusterfutures/compare/poll-slurm...tcompa:clusterfutures:poll-slurm-mod) if you'd like to, but the only real change is adding double quotes around `%i %T`). My version already works as expected within [a PR of ours](https://github.com/fractal-analytics-platform/fractal-server/pull/367), but we subclass/reimplement quite a few components of `clusterfutures` and thus this is not the most straightforward way of testing.

takluyver commented:
Thanks, that sounds good!

Have you found that the extra quotes are needed? I'm using the same squeue option in another tool (sfollow) which I use often, and it seems to be working OK without quotes:

https://github.com/European-XFEL/sfollow/blob/50ec601e622ac971e904a597508b91c91bc135dd/sfollow/sfollow.py#L22-L27

tcompa commented:
> Have you found that the extra quotes are needed? I'm using the same squeue option in another tool (sfollow) which I use often, and it seems to be working OK without quotes:

They are indeed needed, at least on one cluster we use, but it may be due to an old SLURM version:
```
$ squeue  -j 9438391,9435167 --format %i %T -h
squeue: error: Unrecognized option: %T
Usage: squeue [-A account] [--clusters names] [-i seconds] [--job jobid]
              [-n name] [-o format] [-p partitions] [--qos qos]
              [--reservation reservation] [--sort fields] [--start]
              [--step step_id] [-t states] [-u user_name] [--usage]
              [-L licenses] [-w nodes] [-ahjlrsv]

$ squeue  -j 9438391,9435167 --format "%i %T" -h
9438391 RUNNING
9435167 RUNNING

$ sinfo -V
slurm 15.08.7
```

I'll also test a more recent version.

takluyver commented:
The issue there is with writing a command in a shell - it uses spaces to split it into individual arguments, so if you want a  space inside an argument, you need to quote it. Passing a list to `subprocess.run()` avoids this issue, because it's not relying on spaces to split it up. :slightly_smiling_face:

tcompa commented:
> The issue there is with writing a command in a shell - it uses spaces to split it into individual arguments, so if you want a space inside an argument, you need to quote it. Passing a list to `subprocess.run()` avoids this issue, because it's not relying on spaces to split it up. slightly_smiling_face

Thanks, my bad. I now tested your example script from https://github.com/sampsyo/clusterfutures/pull/19#issue-1200192836 directly (after `pip install git+https://github.com/takluyver/clusterfutures.git@poll-slurm`) and it worked - without additional quotes.

---

FYI, the reason why I had added additional quotes is likely due to our CI.

Our pytest CI spins up [a temporary SLURM cluster](https://github.com/rancavil/slurm-cluster) (via [pytest-docker](https://pypi.org/project/pytest-docker/)). Because we have pytest running from outside the container, we need to transform some commands (like `sbatch` or `squeue`) into commands to be executed within the container (e.g. `docker exec CONTAINER_ID bash -c 'sbatch SUBMISSION_SCRIPT'`). I think this transformation is what went wrong in my first try, and that's why I added quotes. But this is something we need to take care of on our side.

takluyver commented:
Thanks, the option to launch a Slurm cluster in Docker may be useful to know about for other testing. :slightly_smiling_face: 

If you need to transform a list of arguments into a shell-style string command for something like `bash -c`, the [`shlex` module](https://docs.python.org/3/library/shlex.html) can do that - `shlex.join(args)` - at least for Unix shells (I don't think there's anything official to do it for Windows, though there's an undocumented function in the subprocess module).

Thanks for testing this. Do you want to do any code review as well, or are you satisfied?

tcompa commented:
> Thanks, the option to launch a Slurm cluster in Docker may be useful to know about for other testing. slightly_smiling_face

We use [these Docker images](https://github.com/fractal-analytics-platform/fractal-server/tree/main/tests/data/slurm_docker_images) and [these pytest fixtures](https://github.com/fractal-analytics-platform/fractal-server/blob/main/tests/fixtures_slurm.py) (ping @mfranzon). We can provide more info if it is useful.

> If you need to transform a list of arguments into a shell-style string command for something like `bash -c`, the [`shlex` module](https://docs.python.org/3/library/shlex.html) can do that - `shlex.join(args)` - at least for Unix shells (I don't think there's anything official to do it for Windows, though there's an undocumented function in the subprocess module).

Thanks! We use it in other places, but probably we missed one ;)

> Thanks for testing this. Do you want to do any code review as well, or are you satisfied?

I'd like to provide a bit more feedback, and I will do it ASAP (probably already today).

tcompa committed:
Add new examples to slurm_example.py

1. What happens in case of an exception;
2. What happens in case of unexpected exit (e.g. a SLURM error).

tcompa committed:
Do not call squeue if there is no job to check

tcompa committed:
Add docstring

tcompa committed:
Add python 3.10 to github CI

sampsyo commented:
So sorry for letting this languish, and thanks for bringing it back up! I'll take a detailed look at the code now.

sampsyo commented on cfut/__init__.py:
```diff
@@ -41,8 +47,7 @@ def __init__(self, callback, interval=1):
         """Stop the thread soon."""
-        with self.lock:
-            self.shutdown = True
+        self.shutdown = True
```
This may be a dumb question: is it safe to eliminate the lock here? `self.shutdown` is still being concurrently modified and read by two different threads, so there is a data race. Maybe this is OK because of the GIL (i.e., writing to the field and reading from it are guaranteed to be atomic because they are single bytecode operations), but it nonetheless makes me a little nervous.

takluyver replied:
I'm pretty confident this is OK. My understanding is that the GIL guarantees that fundamental Python operations happen atomically, and setting an attribute is very fundamental. The glossary entry says:

> This simplifies the CPython implementation by making the object model (including critical built-in types such as [dict](https://docs.python.org/3/library/stdtypes.html#dict)) implicitly safe against concurrent access.

(Typical attribute access like this is based on dict, stored as `obj.__dict__`)

I think we could even get away without it for the `self.waiting` dict, so long as filenames are always unique, but if for any reason a filename was reused, it could get very confusing, so I'll keep it there. And actually, I realise that the extra Slurm checks should also be inside the lock.

sampsyo replied:
Sounds reasonable enough to me!

sampsyo commented on cfut/__init__.py:
```diff
@@ -52,22 +57,27 @@ def wait(self, filename, value):
+            self.check(i)
             time.sleep(self.interval)
 
+    def check(self, i):
```
Perhaps a docstring would be helpful here—I was initially confused as to why `i` was passed as an argument, but I see it now after reading the override in `SlurmWaitThread`.

takluyver replied:
Yup, good idea. :slightly_smiling_face:

sampsyo reviewed (approved):
Fantastic; this looks like a very tidy implementation of the idea. There are two extremely small questions within; I think this is basically good to go.

takluyver committed:
Quote Python versions for Github actions

takluyver committed:
Remove Python 3.6 from tests

takluyver committed:
Merge pull request #1 from tcompa/poll-slurm

Minor changes to https://github.com/sampsyo/clusterfutures/pull/19

takluyver commented:
I merged in @tcompa's changes. We've dropped Python 3.6 from the CI now - it's not available on the new default Ubuntu platform on Github actions. It's easy enough to add a bit of extra config to run 3.6 on the previous Ubuntu if you want to keep supporting it, but this seemed like a reasonable prompt to move on.

takluyver committed:
Acquire lock outside check() to use it for Slurm operations

takluyver committed:
Add docstring for FileWaitThread.check()

sampsyo commented:
It seems like a good idea to drop Python 3.6 from CI, given its EOL status. This looks great to me—merge away! :sparkles:

tcompa commented:
Fantastic, thank you @takluyver!

If you could also make a new PyPI release, that would be great for us (we cannot publish our package to PyPI with a GitHub dependency).

takluyver commented:
No problem, I've just uploaded version 0.5 with this change.

tcompa commented:
Great, thank you
